File: Backtracking_and_Search/LC_51_N_Queens.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.19

class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> ans;
        vector<string> curr;
        dfs(0, n, curr, ans);
        return ans;
    }
private:
    unordered_set<int> used_col, used_diag, used_anti_diag;
    void dfs(int row, int n, vector<string>& curr, vector<vector<string>>& ans){
        if (row == n){
            ans.push_back(curr);
            return;
        }
        for(int col = 0; col < n; col++){
            int diag = row - col, anti_diag = row + col;
            if (used_col.count(col) || used_diag.count(diag) || used_anti_diag.count(anti_diag)){
                continue;
            }
            string temp = string(col, '.') + 'Q' + string(n-col-1, '.');
            used_col.insert(col);
            used_diag.insert(diag);
            used_anti_diag.insert(anti_diag);
            curr.push_back(temp);
            dfs(row + 1, n, curr, ans);
            curr.pop_back();
            used_col.erase(col);
            used_diag.erase(diag);
            used_anti_diag.erase(anti_diag);

        }
    }
};

File: Backtracking_and_Search/LC_51_N_Queens.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.19

class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        used_col, used_diag, used_anti_diag = set(), set(), set()
        ans = []
        
        def dfs(row, curr, ans):
            if row == n:
                ans.append(curr[:])
                return
            for col in range(n):
                diag_idx = row-col
                anti_diag_idx = row + col
                if col in used_col or diag_idx in used_diag or anti_diag_idx in used_anti_diag:
                    continue
                used_col.add(col)
                used_diag.add(diag_idx)
                used_anti_diag.add(anti_diag_idx)
                temp = "." * col + "Q" + "." * (n-col-1)
                curr.append(temp)
                dfs(row+1, curr, ans)
                curr.pop()
                used_col.remove(col)
                used_diag.remove(diag_idx)
                used_anti_diag.remove(anti_diag_idx)

        dfs(0, [], ans)
        return ans